openssl/src/ssl/mod.rs

```diff
@@ -2106,6 +2150,27 @@ impl Ssl {
             }
         }
     }
+
+    /// Perform a stateless handshake
+    ///
+    /// Returns true if a complete ClientHello containing a valid cookie was
+    /// read. Requires that cookie generation and verification callbacks were set on the
+    /// SSL context.
+    ///
+    /// This corresponds to [`SSL_stateless`]
+    ///
+    /// [`SSL_stateless`]: https://www.openssl.org/docs/manmaster/man3/SSL_stateless.html
+    #[cfg(all(feature = "v111", ossl111))]
+    pub fn stateless<S>(&mut self, stream: &mut S) -> bool
```

Comment:
I'd like some input on the design of this. [The docs](https://www.openssl.org/docs/manmaster/man3/SSL_stateless.html) say:

> For SSL_stateless() if an entire ClientHello message cannot be read without the "read" BIO becoming empty then the SSL_stateless() call will fail.

so it seems like it should probably be used with in-memory buffers and always complete immediately? I'm not a huge fan of allocating a new BIO on every call...

Reply:
How is this normally used in C?

Reply:
I'm not sure. The API was only added a few weeks ago in https://github.com/openssl/openssl/pull/4435. The only example I'm aware of is [the unit test](https://github.com/openssl/openssl/blob/master/test/sslapitest.c#L2586), which I don't think tells us much about its intended relationship to async IO.

It's worth noting that in the case of QUIC implementations, [the handshake messages involved will always be sent/received as a single packet](https://quicwg.github.io/base-drafts/draft-ietf-quic-tls.html#clienthello-size), so a fully synchronous paradigm works for my case. I'm also less bothered by the short-lived BIOs than I was, since either the whole `Ssl` object will be disposed immediately regardless (for a ClientHello not containing a valid cookie), or the BIOs only need to be replaced once for the `Ssl`'s entire lifetime, when it's turned into a stream.

Reply:
It's also not apparent to me that this could be used in any way but synchronously, since the interface provides no way to distinguish an incomplete ClientHello from an invalid one.

My understanding is that generally you will either want to dispose of the `Ssl` object entirely if this returns false, or proceed to calling `SSL_accept` if it returns true. That invites tweaking the signature to something like `fn stateless<S>(self, stream: S) -> Option<SslStream<S>>`. I think I'll try to make that work, but I'm not certain that forcing a single inner stream type is convenient.

Reply:
It's intended to be used with datagram sockets, so the streaming issues aren't applicable, right? You always get the whole message in one read call.

I think we may want a lower-level type that lets you drive things more directly. `SSL_read_early_data` is weird in a similar way. Something like

```rust
struct SslStreamBuilder<T>{ ... }

impl<T> SslStreamBuilder<T>
    where T: Read + Write,
{
    pub fn new(stream: T) -> SslStreamBuilder<T> { ... }

    pub fn stateless(&mut self) -> Result<()> { ... }

    pub fn read_early_data(&mut self, buf: &mut [u8]) -> Result<usize> { ... }

    pub fn write_early_data(&mut self, buf: &[u8]) -> Result<usize> { ... }

    pub fn accept(&mut self) -> Result<SslStream<T>, MidHandshakeSslStream<T>> { .. }

    pub fn connect(&mut self) -> Result<SslStream<T>, MidHandshakeSslStream<T>> { .. }
}
```

Reply:
That's an interesting idea, but it places early data I/O *before* the accept/connect call, which seems wrong.

Reply:
> When called by a client, SSL_write_early_data() must be the first IO function called on a new connection, i.e. it must occur before any calls to SSL_write_ex(3), SSL_read_ex(3), SSL_connect(3), SSL_do_handshake(3) or other similar functions.

Reply:
When would you want to recover it? Isn't it going to fail if the client isn't legitimate?

Reply:
`SSL_stateless` "fails" if the client is (potentially) legitimate but does not yet have a cookie, and in doing so replies with a HelloRetryRequest containing a cookie. This needs to actually get sent, or progress will never be made.

Reply:
How would recovering the stream affect SSL_stateless's write being sent or not?

I think we're going to have to look at a real world use of this function to figure out how it's intended to be used.

Reply:
Maybe I'm confused about how this whole wrapped-stream paradigm is supposed to work in general. I don't expect a dropped stream to complete pending writes. In my experimental code, the stream I'm wrapping is just some in-memory buffers, and dropping them would lose the written data entirely.